File: packages/sync-client/src/validation.ts
import { z } from 'zod';

export const ConversationSyncPayloadSchema = z.object({
  id: z.number().optional(),
  local_id: z.string().optional(),
  timestamp: z.string(),
  user_id: z.string().optional(),
  user_input: z.string(),
  result: z.string().optional(),
  execution_time: z.number().optional(),
  model: z.string().optional(),
  agent_count: z.number().optional(),
  sync_version: z.number(),
  last_synced_at: z.string(),
  device_id: z.string().optional(),
  is_deleted: z.boolean(),
  is_archived: z.boolean().optional(),
  content_truncated: z.boolean().optional(),
  updated_at: z.string(),
});

export const MessageSyncPayloadSchema = z.object({
  message_id: z.string(),
  conversation_id: z.number(),
  conversation_local_id: z.string().optional(),
  role: z.string(),
  content: z.string(),
  is_streaming: z.boolean(),
  is_agent_status: z.boolean(),
  elapsed_seconds: z.number().optional(),
  created_at: z.string(),
  error: z.string().optional(),
  sources: z.unknown().optional(),
  tool_events: z.unknown().optional(),
  agent_statuses: z.unknown().optional(),
  trace: z.unknown().optional(),
  sync_version: z.number(),
  last_synced_at: z.string(),
  device_id: z.string().optional(),
  is_deleted: z.boolean(),
  content_truncated: z.boolean().optional(),
  updated_at: z.string(),
});

export const DeletionRecordSchema = z.object({
  type: z.enum(['conversation', 'message']),
  id: z.string(),
  deleted_at: z.string(),
});

export const SyncPullResponseSchema = z.object({
  conversations: z.array(ConversationSyncPayloadSchema),
  messages: z.array(MessageSyncPayloadSchema),
  deletions: z.array(DeletionRecordSchema),
  latest_version: z.number(),
  has_more: z.boolean().optional(),
  state_hash: z.string().optional(),
});

export const ConflictRecordSchema = z.object({
  type: z.enum(['conversation', 'message']),
  id: z.string(),
  reason: z.string(),
  server_version: z.number(),
  client_version: z.number(),
});

export const SyncPushResponseSchema = z.object({
  accepted: z.array(z.string()),
  conflicts: z.array(ConflictRecordSchema),
  new_version: z.number(),
  conversation_id_mappings: z.record(z.string(), z.number()),
});

export const SyncStatusResponseSchema = z.object({
  last_synced_at: z.string().optional().default(''),
  sync_version: z.number(),
  pending_changes: z.number(),
});

export const TokenResponseSchema = z.object({
  token: z.string().optional(),
  expires_in: z.number().positive().optional(),
});

export const ErrorStatusSchema = z.object({
  status: z.number().optional(),
});

File: packages/adapters/pkg/sync/validation.go
package sync

import (
	"errors"
)

// Validatable defines an interface for structs that can validate themselves.
type Validatable interface {
	Validate() error
}

func (e BroadcastEvent) Validate() error {
	if e.Type == "" {
		return errors.New("type is required")
	}
	return nil
}

func (c ConversationSyncPayload) Validate() error {
	if c.Timestamp == "" {
		return errors.New("timestamp is required")
	}
	if !c.IsDeleted && c.UserInput == "" {
		return errors.New("userInput is required")
	}
	if c.UpdatedAt == "" {
		return errors.New("updatedAt is required")
	}
	return nil
}

func (m MessageSyncPayload) Validate() error {
	if m.MessageID == "" {
		return errors.New("messageId is required")
	}
	if m.Role == "" {
		return errors.New("role is required")
	}
	if !m.IsDeleted && m.Content == "" {
		return errors.New("content is required")
	}
	if m.CreatedAt == "" {
		return errors.New("createdAt is required")
	}
	if m.UpdatedAt == "" {
		return errors.New("updatedAt is required")
	}
	return nil
}

func (d DeletionRecord) Validate() error {
	if d.Type != "conversation" && d.Type != "message" {
		return errors.New("invalid deletion type")
	}
	if d.ID == "" {
		return errors.New("id is required")
	}
	if d.DeletedAt == "" {
		return errors.New("deletedAt is required")
	}
	return nil
}

func (r SyncPullResponse) Validate() error {
	for _, c := range r.Conversations {
		if err := c.Validate(); err != nil {
			return err
		}
	}
	for _, m := range r.Messages {
		if err := m.Validate(); err != nil {
			return err
		}
	}
	for _, d := range r.Deletions {
		if err := d.Validate(); err != nil {
			return err
		}
	}
	return nil
}
